add marks section to clay docs

tinnus-napbus opened the pull request:
- add overview
- add writing marks guide
- add using marks guide
- replace marks section of clay architecture doc with a link to the new
  marks section
- fix up links etc

tinnus-napbus committed:
add marks section to clay docs
- add overview
- add writing marks guide
- add using marks guide
- replace marks section of clay architecture doc with a link to the new
  marks section
- fix up links etc

drbeefsupreme commented on content/docs/arvo/clay/marks/writing-marks.md:
```diff
@@ -0,0 +1,331 @@
+
+There is a little complexity surrounding special characters in fields and line endings, but otherwise the only other rule is that all rows must have the same number of fields. You can refer to [RFC4180 on the IETF website](https://datatracker.ietf.org/doc/html/rfc4180) for more details.
+
+The simplest way to represent such a structure in urbit is as a `(list (list @t))` like:
```
```suggestion
The simplest way to represent such a structure in Hoon is as a `(list (list @t))` like:
```

drbeefsupreme commented on content/docs/arvo/clay/marks/marks.md:
```diff
@@ -0,0 +1,57 @@
+template = "doc.html"
++++
+
+Clay is a typed filesystem, and we call these file types `mark`s. When talking about Hoon and Arvo we'll often talk of types like `@ud`, `(list @t)`, etc. A `mark` will specify such a type for its files, but it does more than just that - it also defines conversion routines to and from other `mark`s, as well as diff, patch, and merge routines.
```
Somewhere on this page you should mention that marks are located in the `/mar` folder

drbeefsupreme commented on content/docs/arvo/clay/marks/writing-marks.md:
```diff
@@ -0,0 +1,331 @@
+
+So with the nature of the `%mime` `mark` hopefully now clear, the reason we want conversion methods to and from `%mime` in our `%csv` `mark` is so we can import CSV files from Unix and vice versa.
+
+Since a CSV file on Unix will just be a long string with ASCII or UTF-8 encoding, we can treat `q.q` in the `$mime` as a `cord`, and thus write a parser to convert it to a `(list (list @t))`. For this, I've created a `csv.hoon` library, which you can view in full on the [Examples](/docs/arvo/clay/marks/examples#libcsvhoon) page.
```
"For this, I've created" First person sounds weird in documentation, maybe "We have a `csv.hoon` library...`?

drbeefsupreme commented on content/docs/arvo/clay/marks/examples.md:
````diff
@@ -0,0 +1,248 @@
+
+```hoon
+|%
+++  validate                                             ::  All rows same length?
````
This is more of a side comment than a suggestion that you may have already considered, but it could potentially simplify some things to use `$|` to get a more restricted `(list (list @t))` type where all rows are the same length rather than calling `validate` in several locations. I haven't actually worked this through yet, and there might be pedagogical or stylistic reasons to use `(list (list @t))` rather than `$|  (list (list @t))  validate`, but perhaps it could be included as an exercise at the end if it actually makes sense to do so.

tinnus-napbus replied:
oh true I wasn't familiar with `$|` tbh, I'll take a look and see if I can change it to using that

drbeefsupreme commented on content/docs/arvo/clay/marks/examples.md:
```diff
@@ -0,0 +1,248 @@
+
+These are the files used in the [Writing Marks](/docs/arvo/clay/marks/writing-marks) guide.
+
+## `/lib/csv/hoon`
```
There's already a `/lib/csv/hoon` https://github.com/urbit/urbit/blob/master/pkg/arvo/lib/csv.hoon so this is probably going to confuse the reader without some clarification

tinnus-napbus replied:
dammit I never noticed that existed. I'll change the name of the library and add a clarifying note

drbeefsupreme commented on content/docs/arvo/clay/marks/writing-marks.md:
```diff
@@ -0,0 +1,331 @@
+
+So far we've just delegated `+grad` functions to the `%noun` `mark`, but now we'll look at writing our own.
+
+For demonstrative purposes I've just poached the algorithms used in the `+grad` arm of the `%txt` `mark` and modified them to take our `(list (list @t))` type instead of a `wain`. It's not the most efficient algorithm for a CSV file but it'll do the job.
```
Another use of first person

drbeefsupreme commented on content/docs/arvo/clay/marks/using-marks.md:
```diff
@@ -0,0 +1,240 @@
+template = "doc.html"
++++
+
+In the last document, [Writing Marks](/docs/arvo/clay/marks/writing-marks), we walked through writing `mark` files and touched on how Clay handles them. They needn't just be left to background Vane processes though, you can also use them yourself in your code.
```
```suggestion
In the last document, [Writing Marks](/docs/arvo/clay/marks/writing-marks), we walked through writing `mark` files and touched on how Clay handles them. They needn't just be left to background vane processes though, you can also use them yourself in your code.
```

drbeefsupreme reviewed (changes requested):
Another great PR. Only some minor comments.

tinnus-napbus committed:
clay: marks: using-marks: fix minor typo

Co-authored-by: poprox <51337059+drbeefsupreme@users.noreply.github.com>

tinnus-napbus committed:
clay: marks: writing-marks: minor fix

Co-authored-by: poprox <51337059+drbeefsupreme@users.noreply.github.com>

tinnus-napbus committed:
clay: marks: change lib name and 1st person refs
- change csv.hoon lib to csv-mark.hoon so as not to conflict with
  existing lib
- add note about the distinction
- rephrase 1st person parts

tinnus-napbus committed:
clay: marks: add note about /mar dir

tinnus-napbus committed:
add note about $| rune possible alternative

tinnus-napbus commented:
@drbeefsupreme I have addressed the points you raised. Wrt the point about possibly using `$|`, I played around with it and I think it's simpler for demonstrative purposes to just use a `(list (list @t))`. Because the validation function in the `$|` is only called when you clam a noun as opposed to when you cast something, it would still require strategically clamming things similar to how I've put a `?>` in various places. So it's not all that much less complex while also having to define the type itself in the separate library and thus more complex in terms of how the walk-through proceeds.

Also master-morzod seemed to think it would be acceptable to do it this way

I have however added a note that you could use a `$|` instead and that we're doing it this way for the sake of demonstrative simplicity

drbeefsupreme commented:
LGTM! tagging @philipcmonk or @belisarius222 to check technical aspects

tinnus-napbus committed:
change csv lib name and add note about name
- change csv lib name back to csv.hoon
- add note about the mark and lib conflicting with existing files

tinnus-napbus commented:
ok `|crunch` introduced a new csv mark that conflicts with the example used in this doc. Since both the mark and lib now conflict with existing files, I've just changed them both to csv.hoon and added a note about how they conflict and that this isn't a problem on a fakezod

tinnus-napbus committed:
remove redundant comment re: csv lib

drbeefsupreme commented:
#1060 should be closed once this is merged

tinnus-napbus committed:
clay/marks overview: clarify +grow is optional

tinnus-napbus committed:
clay/marks writing: clarify +grow arm not mandatory

drbeefsupreme commented:
Still looking to get this checked @philipcmonk @belisarius222
